### PR TITLE
[CvT] Tensorflow implementation

### PR DESCRIPTION
**Pull request by mathieujouffroy**
# What does this PR do?
This PR adds the Cvt model implementation in Tensorflow.
This includes the base model and the model with an image classification head on top.

## TODO
- [x] Write the fundamental components (Convolutional Token Embeddings & Convolutional Transformer Block)
- [x] Write base model & image classification model 
- [x] Modify related utilities 
- [x] Write relevant tests (in test suite)
- [x] Preview Tensorflow documentation for Cvt 

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

### Questions
- In the configuration file of the model CVT, ```layer_norm_eps``` is initialized at ```1e-12```.
However, it seems that in the original implementation, the authors use ```epsilon=1e-5```.
Moreover, the Cvt model in pytorch (HuggingFace), does not seem to use the configuration ```layer_norm_eps=1e-12``` for layer normalization throughout the model, instead using the default ```epsilon=1e-5```.
What is the use of layer_norm_eps in the configuration file (of the Cvt model) ? 

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by LysandreJik** [comment]
Thanks for your PR @mathieujouffroy! Let me ping @amyeroberts for review :)

**Comment by mathieujouffroy** [comment]
You're welcome. Cool thanks, should I create an Issue ?

File: src/transformers/models/cvt/modeling_tf_cvt.py
Comment (amyeroberts) [inline src/transformers/models/cvt/modeling_tf_cvt.py:322]: nit: Can you give more descriptive names to the vars here?

File: src/transformers/models/cvt/modeling_tf_cvt.py
Comment (amyeroberts) [inline src/transformers/models/cvt/modeling_tf_cvt.py:346]: nit: Same here re descriptive var names

File: src/transformers/models/cvt/modeling_tf_cvt.py
Comment (amyeroberts) [inline src/transformers/models/cvt/modeling_tf_cvt.py:323]:
```suggestion
        batch_size, hidden_size, num_channels = tf.shape(hidden_state)
```

File: src/transformers/models/cvt/modeling_tf_cvt.py
Comment (amyeroberts) [inline src/transformers/models/cvt/modeling_tf_cvt.py:347]:
```suggestion
        _, _, hidden_size, _ = tf.shape(context)
```

File: src/transformers/models/cvt/modeling_tf_cvt.py
Comment (amyeroberts) [inline src/transformers/models/cvt/modeling_tf_cvt.py:84]: nit: Rather than have most of the logic indented, you can rewrite for an early return condition

```suggestion
        if not training:
            return x
        keep_prob = 1 - self.drop_prob
        shape = (tf.shape(x)[0],) + (1,) * (len(tf.shape(x)) - 1)
        random_tensor = keep_prob + tf.random.uniform(shape, 0, 1)
        random_tensor = tf.floor(random_tensor)
        return (x / keep_prob) * random_tensor
```

File: src/transformers/models/cvt/modeling_tf_cvt.py
Comment (amyeroberts) [inline src/transformers/models/cvt/modeling_tf_cvt.py:587]: nit: More descriptive var names in the comment please 🙏

File: src/transformers/models/cvt/modeling_tf_cvt.py
Comment (amyeroberts) [inline src/transformers/models/cvt/modeling_tf_cvt.py:603]: nit: More descriptive var names please :)

File: src/transformers/models/cvt/modeling_tf_cvt.py
Comment (amyeroberts) [inline src/transformers/models/cvt/modeling_tf_cvt.py:928]: nit: More descriptive var names please

File: src/transformers/models/cvt/modeling_tf_cvt.py
Comment (amyeroberts) [inline src/transformers/models/cvt/modeling_tf_cvt.py:929]:
```suggestion
            batch_size, num_channels, height, width = tf.shape(sequence_output)
```

File: src/transformers/models/cvt/modeling_tf_cvt.py
Comment (amyeroberts) [inline src/transformers/models/cvt/modeling_tf_cvt.py:586]:
```suggestion
        batch_size, num_channels, height, width = tf.shape(hidden_state)
```

**Review by amyeroberts — APPROVED** [review]
Thanks for adding the model and for such a nice PR ❤️ Just a few nits and one comment regarding a permutation. 

We recently had the first community contributor open a PR to add weights on the hub (see [PR comment here on steps](https://github.com/huggingface/transformers/pull/18678#issuecomment-1222244001)). Once we've had two other 👍  we should be good to add the model weights, run the slow tests,  and then merge.

File: src/transformers/models/cvt/modeling_tf_cvt.py
Comment (mathieujouffroy) [reply src/transformers/models/cvt/modeling_tf_cvt.py:229]: Oh yess sorry thanks. This is due to me using ```tf.keras.layers.DepthwiseConv2D```, the pytorch implementation uses `nn.Conv2d` with the `groups` parameter. However, if I use the permutation `NCHW` to `NHWC` with ```tf.keras.layers.DepthwiseConv2D```,  I get  a difference in the output tensors. Even though the tensorflow model and the pytorch model predict the same class, the difference in their output tensors is relatively big (thus failing the inference test).

I would like like to have your advice on the preposed improvements : 
- I can change the comment and leave the permutation as so -> `NCHW` to `NWHC`.
- I can update the permutations accordingly to the comments  (`NCHW` to `NHWC`) and use `tf.keras.layers.Conv2D`with the `groups` parameter. This modification will also avoid writing the condition I put in the function `convert_tf_weight_name_to_pt_weight_name` (in the file `modeling_tf_pytorch_utils.py`).

File: src/transformers/models/cvt/modeling_tf_cvt.py
Comment (mathieujouffroy) [reply src/transformers/models/cvt/modeling_tf_cvt.py:156]: No indeed it's never False, thks 😊

File: src/transformers/models/cvt/modeling_tf_cvt.py
Comment (mathieujouffroy) [reply src/transformers/models/cvt/modeling_tf_cvt.py:153]: I tried using ```tf.shape``` but unfortunately it does break things when running the tests 😕.  I get an error : ```OperatorNotAllowedInGraphError: Iterating over a symbolic `tf.Tensor` is not allowed in Graph execution```
I could retrieve the individual shapes differently though, but I guess that using ```shape_list``` is more consistent. Should I leave it like that ?

File: src/transformers/modeling_tf_pytorch_utils.py
Comment (gante) [inline src/transformers/modeling_tf_pytorch_utils.py:94]: If we don't make this change, what is the error that pops up? 

(If possible, I'd like to handle the issue in the modeling file rather than to these shared functions :) )

File: src/transformers/models/cvt/modeling_tf_cvt.py
Comment (gante) [inline src/transformers/models/cvt/modeling_tf_cvt.py:60]: Looking at the docstring, I can see that the output shapes do not match their PT counterpart. Assuming PT's docstrings are correct, we need to change the outputs here (and their docstrings) to match PT's -- keeping a consistent API is very important 🙏

File: src/transformers/models/cvt/modeling_tf_cvt.py
Comment (gante) [reply src/transformers/models/cvt/modeling_tf_cvt.py:84]: The shortcut return path can also have `or self.drop_prob == 0.0` 🚤

File: src/transformers/models/cvt/modeling_tf_cvt.py
Comment (gante) [inline src/transformers/models/cvt/modeling_tf_cvt.py:111]: missing `training` argument (for the dropout)

It is also missing in other layers -- don't forget to add the argument on upstream layers 👍

File: src/transformers/models/cvt/modeling_tf_cvt.py
Comment (gante) [inline src/transformers/models/cvt/modeling_tf_cvt.py:256]: The `qkv_bias` argument is missing here (and in other classes), and then in the initialization of the Dense layers

**Review by gante — COMMENTED** [review]
Thank you for the addition of CvT 🔥 

I've added a few comments in the code, the most concerning one being the comment about the output shapes.

After the issues raised by me and @amyeroberts, we can finalize the review process by:
1. tagging a core maintainer for a third review;
2. opening PRs for the TF weights (I will share the instructions on how to do it :) )

File: src/transformers/modeling_tf_pytorch_utils.py
Comment (mathieujouffroy) [reply src/transformers/modeling_tf_pytorch_utils.py]: Hi, I corrected the problem directly in the modeling file 😊. This condition was to avoid having weight initialization problems. In the tensorflow implementation I was using `tf.keras.layers.DepthwiseConv2D` instead of `tf.keras.layers.Conv2D` with the `groups` parameter. Therefore the tensorflow weight name was different than the pytorch weight name.

File: src/transformers/models/cvt/modeling_tf_cvt.py
Comment (amyeroberts) [reply src/transformers/models/cvt/modeling_tf_cvt.py]: Yes, unfortunately depthwise convolutions are a bit tricky :( 

As you pointed out, `Conv2D` with the groups params can be used instead. This is what's been done in other TF implementations e.g. [in Segformer](https://github.com/huggingface/transformers/blob/142e12afb457e31ec0e7323d9eba7b8bfee46c57/src/transformers/models/segformer/modeling_tf_segformer.py#L243). If it's possible to apply here, then that's what I'd recommend.

**Review by gante — APPROVED** [review]

File: src/transformers/models/cvt/modeling_tf_cvt.py
Comment (mathieujouffroy) [reply src/transformers/models/cvt/modeling_tf_cvt.py]: Okay perfect thanks. I corrected it by using `tf.keras.layers.Conv2D`

**Comment by mathieujouffroy** [comment]
Thanks a lot for both of your reviews 🙏 ! 
I've corrected the issues :) 
Although, I kept using `shape_list` instead of `tf.shape` throughout the implementation of the model as `tf.shape` was breaking things while running the tests (see comment above).
Should I follow the instructions in this [PR comment](https://github.com/huggingface/transformers/pull/18678#issuecomment-1222244001) to upload to weights ?

**Comment by gante** [comment]
@mathieujouffroy awesome, seems like we are ready to move on to the next stage. I'm adding @sgugger as the last reviewer.

Meanwhile, you can open the PR to the TF model weights on the hub as follows:
1. Make sure you have the latest version of the hub installed (`pip install huggingface_hub -U`) and that you are logged in to HF with a write token (`huggingface-cli login`)
2. Run `transformers-cli pt-to-tf --model-name foo/bar` from this branch :D
3. In the Hub PR, tag `@joaogante, @nielsr, @sgugger`

**Review by sgugger — APPROVED** [review]
Thanks a lot for your PR! Good to merge once the weights have been added!

**Comment by mathieujouffroy** [comment]
> @mathieujouffroy awesome, seems like we are ready to move on to the next stage. I'm adding @sgugger as the last reviewer.
> 
> Meanwhile, you can open the PR to the TF model weights on the hub as follows:
> 
> 1. Make sure you have the latest version of the hub installed (`pip install huggingface_hub -U`) and that you are logged in to HF with a write token (`huggingface-cli login`)
> 2. Run `transformers-cli pt-to-tf --model-name foo/bar` from this branch :D
> 3. In the Hub PR, tag `@joaogante, @nielsr, @sgugger`

I am getting an error when using `transformers-cli pt-to-tf --model-name microsoft/cvt-13` : 
```
File "/Users/MathieuJouffroy/transformers/src/transformers/commands/pt_to_tf.py", line 307, in run
    + "\n".join([f"{k}: {v:.3e}" for k, v in hidden_differences.items() if v > self._max_error])
ValueError: The cross-loaded TensorFlow model has different outputs, something went wrong!

List of maximum output differences above the threshold (5e-05):
logits: 1.190e-04

List of maximum hidden layer differences above the threshold (5e-05):
hidden_states[2]: 1.227e-02
```

It seems that both the `max_crossload_output_diff `and the `max_crossload_hidden_diff` are bigger than the `self._max_error` **(5e-5)**. 
Respectively I have `max_crossload_output_diff` = 0.00011897087 **(1.190e-04)** and  `max_crossload_hidden_diff` = 0.012268066 **(1.227e-02)**.

I am trying to figure out how to correct this error (WIP).

**Comment by gante** [comment]
@mathieujouffroy ~1e-2 is quite large -- does this happen exclusively on `microsoft/cvt-13`, or across all CvT models?

**Comment by mathieujouffroy** [comment]
> @mathieujouffroy ~1e-2 is quite large -- does this happen exclusively on `microsoft/cvt-13`, or across all CvT models?

Yess, unfortunately it happens across all CvT models.
When inspecting the difference between the hidden states of the pytorch model and the hidden states of the tensorflow model, I can see that the difference increases throughout the model (with the number of layers).
The CvT model is composed of 3 stages of encoder block, with their respective number of layers being 1, 2 and 10. In the last stage, the difference between the torch model's hidden state and the tensorflow model's hidden state increases from ~e-5 at layer[0] to ~1e-2 at layer[9].
For `microsoft/cvt-13`  : 
```
CvtLayer output vs TFCvtLayer output

diff pt-tf stage[0]/layer[0]: 1.77919864654541e-05

diff pt-tf stage[1]/layer[0]: 2.4199485778808594e-05
diff pt-tf stage[1]/layer[1]: 3.9249658584594727e-05

diff pt-tf stage[2]/layer[0]: 3.0934810638427734e-05
diff pt-tf stage[2]/layer[1]: 0.000102996826171875
diff pt-tf stage[2]/layer[2]: 0.0004825592041015625
diff pt-tf stage[2]/layer[3]: 0.0009307861328125
diff pt-tf stage[2]/layer[4]: 0.001621246337890625
diff pt-tf stage[2]/layer[5]: 0.0032196044921875
diff pt-tf stage[2]/layer[6]: 0.0064239501953125
diff pt-tf stage[2]/layer[7]: 0.0091705322265625
diff pt-tf stage[2]/layer[8]: 0.012481689453125
diff pt-tf stage[2]/layer[9]: 0.01226806640625

Hidden Differences:
hidden_states[0]:1.77919864654541e-05
hidden_states[1]:3.9249658584594727e-05
hidden_states[2]:0.01226806640625

Output Differences:
logits:0.00011897087097167969
```
I can't seem to correct this issue. I was wondering if this was due to floating points operations.
Do you have any advice ? 🙏

**Comment by gante** [comment]
@mattchurgin in these cases, a deep dive has to be done -- place a pair of `breakpoint()` in the layer where the problems start, one in each framework, and see which operation causes the divergence. Then, confirm that the TF operation/layer is parametrized correctly and, if it is, one has to dig even deeper :D

**Comment by gante** [comment]
@mathieujouffroy That's a great dive in! 

Previously we didn't have these checks in place, so it is possible that issues like the one you're seeing slipped through the cracks. It's not positive at all to have such large mismatches. I've had in my plans to go back and double-check the most popular models with the recently introduced checks, and you've just raised the priority of the task with your message :)

I think @amyeroberts has seen similar TF/PT mismatches due to the Batch Normalization layer. @amyeroberts do you mind pitching in?

**Comment by amyeroberts** [comment]
@mathieujouffroy Thanks for all the work digging into this 🕵️  

As `momentum` is set for both the pytorch and TF models, I believe their behaviour (outputs and moving stats updates) _should_ be the same during both inference and training, given the same weights and params.  

@gante @mathieujouffroy Yes, I had similar issues with the TF ResNet port ([a weights PR for reference](https://huggingface.co/microsoft/resnet-152/discussions/1)). Like this model, the batch norm layer introduced differences which then got amplified through the forward pass. @ydshieh did some excellent detective work, and found that matching all of the parameters and inputs to produce an equivalent TF and PyTorch layer would still produce outputs with a difference on the order of `1e-7` (enough to start causing problems 😭)  

Ultimately, we decided to add the weights as the difference between the logits was small ~1e-5. I think the ~1e-4 absolute differences in this case are acceptable for adding the weights. @sgugger Is this OK?

**Comment by sgugger** [comment]
Yes, as long as it stays in the range of 1e-4, we can accept the difference between frameworks.

**Comment by gante** [comment]
Thank you for pitching in @amyeroberts :D

@mathieujouffroy feel free to use `--max-error 1e-4` (or slightly higher) in the `pt-to-tf` CLI to ignore those errors and push the weights!

**Comment by mathieujouffroy** [comment]
@gante @amyeroberts you're welcome and thanks a lot for your feedbacks 😊 !

@amyeroberts It seems that for the batch normalization, the update rule for the running statistics is slightly different in Tensorflow compared to Pytorch:
PT -> `running_mean = (1 - momentum) * running_mean + momentum * new_value`
TF -> `running_mean = momentum * running_mean  +  (1 - momentum) * new_value`
Therefore, I think I made a mistake in setting the momentum to 0.1 in TF Batch Norm. Considering the update rules, shouldn't the momentum be set to 0.9 (default) in TF when it is set to 0.1 (default) in PT ? 
However, even though I change the momentum, I still have the same difference in my outputs 😕.

@gante Ok thanks. Although, just as a reminder, when I keep the Batch Normalization layers, I have a `max_crossload_output_diff` of ~1e4 and a `max_crossload_hidden_diff` of  ~2e-2 for all CvT models except the `cvt-21-384-22k`. The `cvt-21-384-22k` has a `max_crossload_output_diff` of ~4e4 and a `max_crossload_hidden_diff` of  ~1e-1.
Therefore, should I use `--max-error 1e-2` for all CvT models and  `--max-error 1e-1` for `cvt-21-384-22k` ? 

I'll also be more than happy to help if you need any assistance regarding the mismatches between PT and TF (I'm a bit intrigued) 😊

**Comment by mathieujouffroy** [comment]
Hello @gante @amyeroberts, as pointed out in this [PR](https://github.com/huggingface/transformers/pull/19341),  the dense layer weights for PT should be initialized with `nn.init.trunc_normal_` instead of `normal_` as in the original implementation of the Cvt model (which uses `trunc_normal_` from `timm` library). In TF `get_initializer` already returns `tf.keras.initializers.TruncatedNormal`.
Also, following the original implementation,  in both frameworks the `cls_token` should be initialized with `trunc_normal` as with the `config.initializer_range` (here 0.02).

Should I add the modifications regarding the `momentum` (setting it to 0.9 in TF) and the use of `trunc_normal` ?

**Comment by gante** [comment]
@mathieujouffroy regarding initialization yeah, update it if possible :) In any case, it has come to our attention that TF initialization is very wrong in some places, and we will be adding tests and updates in the coming weeks!

Regarding momentum, I will defer the decision to @amyeroberts, who has been working more closely with that layer.

**Comment by amyeroberts** [comment]
@mathieujouffroy Thanks for the update. 

Regarding the initialisation, could you update the PyTorch model in a separate PR and do similar thing as [suggested here](https://github.com/huggingface/transformers/pull/19341#pullrequestreview-1131874527) - naming the PR with a 🚨🚨🚨 prefix so we can easily spot and flag in the release notes. 

For momentum in the batch norm layers, yes please use the (1 - momentum) value :)

**Comment by mathieujouffroy** [comment]
@gante @amyeroberts Okay thanks, I will update the changes concerning the TF model (cls_token initialization & momentum) in this PR and will create a new PR for the Pytorch model 😊.

What should I do regarding the use of `pt-to-tf` (and `--max-error`) to add the weights ?
@gante should I wait for the future tests and updates you were mentioning regarding the TF initialization ?
Or should I add the weights with the `--max-error` I mentioned in the comment above ? 
Thanks !

**Comment by gante** [comment]
@mathieujouffroy the initialization shouldn't change the mistmatch between the two frameworks for pretrained models -- I'd recommend going forward with `--max-error` 💪

**Comment by mathieujouffroy** [comment]
Hi @amyeroberts, thanks for your mention !! I've added the [PR](https://github.com/huggingface/transformers/pull/19486) regarding the pytorch model. 

@gante following your recommendation I've added the weights on the hub 😊
As @amyeroberts had pointed out, I'll need to remove `from_pt` in the testing file once the weights are added.

**Comment by gante** [comment]
@mathieujouffroy weights merged 🙌

**Comment by mathieujouffroy** [comment]
> @mathieujouffroy weights merged 🙌

Cool thanks @gante 😊 !
I'll update the testing file.

**Comment by gante** [comment]
@mathieujouffroy off-topic: are you working with `transformers` as part of École 42? I've been at the school once (like 5+ years ago) and I had a friend who participated -- I really liked the concept!

**Comment by mathieujouffroy** [comment]
> @mathieujouffroy off-topic: are you working with `transformers` as part of École 42? I've been at the school once (like 5+ years ago) and I had a friend who participated -- I really liked the concept!

@gante Yess I was working with `transformers` 🤗 on my last project (computer vision) at École 42. The project was in partnership with Hectar, an agricultural campus. I was pretty excited to try out the vision transformers 😊.
I've also used `transformers` at 42 for my NLP projects and in my internship (in NLP). 
I thinks 42 is a really cool training 🚀 : project-based & peer to peer pedagogy !

File: tests/models/cvt/test_modeling_tf_cvt.py
Comment (gante) [reply tests/models/cvt/test_modeling_tf_cvt.py:257]: (@mathieujouffroy FYI I've updated this line)

**Comment by gante** [comment]
All seems ready, merging as soon as CI turns green.

@mathieujouffroy on behalf of TF users, thank you for making the ecosystem richer 🧡

**Comment by mathieujouffroy** [comment]
@gante @amyeroberts thanks a lot for your help and feedbacks !! 💛
It was really interesting and cool to do this PR (1st in an open source project) and to get it merge 😊

**Pull request merged**
